port-forward listen on address

goblain opened the pull request:
**What this PR does / why we need it**:

Implements #43962 proposal. Adds `--address` flag to port-forward command that allows listening on addresses other then localhost, so that port-forward can ie. be opened to consumers other then residing in local host like running in docker or different machine/vm

**Which issue this PR fixes**: 

fixes #43962, fixes #36152, fixes #29678

**Release note**:
```
allows selecting non-localhost addresses to listen on with port-forward
```

mengqiy commented:
@goblain Are you still working on this? If so, please rebase. I can help review this.

mengqiy commented:
/assign

goblain commented:
@mengqiy please take a look at this

mengqiy commented:
/ok-to-test
@goblain Thanks for contributing.
Can you add some test please?

mengqiy commented:
/retest

goblain commented:
/retest pull-kubernetes-unit

mengqiy commented:
/retest

mengqiy commented:
@goblain Any update on the failing test?

goblain commented:
/test pull-kubernetes-bazel-test

goblain commented:
@mengqiy : sorry for the lag, but I finally refactored this PR to support testing that is in line with existing tests and also added ability to provide multiple comma separated addresses to listen on (including fqdn that will be resolved to the list of IPs to try to listen on)

mengqiy commented:
Yay, all tests are passing now. I will do a pass soon.

mengqiy commented on staging/src/k8s.io/client-go/tools/portforward/portforward.go:
```diff
@@ -110,8 +111,68 @@ func parsePorts(ports []string) ([]ForwardedPort, error) {
+	// assemble and return []listenAddress to listen on
+	for _, address := range addressesToParse {
+		if net.ParseIP(address).To4() != nil {
+			parsed = []listenAddress{{address: address, protocol: "tcp4"}}
```
It seems `parsed` will always have one item, why make it a list?

goblain replied:
you're completely right, leftover from some refactoring

mengqiy commented:
/unassign jlowdermilk davidopp

mengqiy commented:
/assign @liggitt
PTAL at client-go part.

liggitt commented:
@smarterclayton what’s the status of switching portforward to websockets from spdy? This seems likely to broaden use of the portforward feature… will the switch involve API level changes we want to make before encouraging broader usage?

smarterclayton commented:
Yes, we're going to have to introduce a new kubelet to master to client
protocol for web sockets that can handle stream framing in order to handle
multiple connections being forwarded.  I'd like to avoid anything impactful
to port forward until we make that change.

On Thu, Sep 21, 2017 at 11:10 PM, Jordan Liggitt <notifications@github.com>
wrote:

> @smarterclayton <https://github.com/smarterclayton> what’s the status of
> switching portforward to websockets from spdy? This seems likely to broaden
> use of the portforward feature… will the switch involve API level changes
> we want to make before encouraging broader usage?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/46517#issuecomment-331338650>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p7uywPAWTshDBA49rX_lXccWKbrtks5skyUXgaJpZM4Nny91>
> .
>

jeffmendoza commented:
@liggitt @smarterclayton This would be a useful feature to kubectl, and updates?

forelabs commented:
@goblain please rebase :)

otherwise if nothing will happen, how about do a fork , get those commits and make a new pr?
I rly think this feature is useful.

goblain commented:
yea, well it kind of got put on hold by expected changes in client, no idea if it's ok to go with it again
@smarterclayton ?

chadnickbok commented:
Its frustrating that this very reasonable change has been left to die on the vine for months because... No reason really? Some other work was happening and some dude got nervous and decided not to add this right now?

This feature is great and should be evaluated on its merits; as far as I can tell the switch from SPDY to websockets won't impact this code at all.

jeffmendoza commented:
Bump  @kubernetes/sig-cli-pr-reviews

goblain commented:
rebased

mengqiy commented:
> Yes, we're going to have to introduce a new kubelet to master to client
protocol for web sockets that can handle stream framing in order to handle
multiple connections being forwarded.  I'd like to avoid anything impactful
to port forward until we make that change.

@smarterclayton @liggitt What is the status now? Can we merge this? It seems people really like this feature.

mengqiy commented:
cc: @seans3 @soltysh

mrnicoco commented:
it's going to be a year now, that's sad. What's missing to be merged ?

liggitt commented on staging/src/k8s.io/client-go/tools/portforward/portforward.go:
```diff
@@ -110,8 +111,68 @@ func parsePorts(ports []string) ([]ForwardedPort, error) {
+			al = append(al, addr)
+		} else {
+			var hosts []string
+			hosts, err := net.LookupHost(addr)
```
I wouldn't expect to do DNS resolution internally... specifying the actual address (`127.0.0.1`, `::1`, `1.2.3.4`, `0.0.0.0`, etc) is consistent with all other bind address options in kubernetes commands

goblain replied:
I must have had a reason to include it, but after all this time I don't really remember. Probably just some minor convenience like not having to script IP4/6 retrieval when you know the name. I am not going to argue much in favor of it though, so if that is a problem, I'll just strip it away.

lavalamp replied:
(FTR, I interpret @liggitt's comment as meaning "don't accept localhost as an option, so that resolution isn't necessary".)

liggitt commented on pkg/kubectl/cmd/portforward.go:
```diff
@@ -109,6 +116,7 @@ func NewCmdPortForward(f cmdutil.Factory, cmdOut, cmdErr io.Writer) *cobra.Comma
 		},
 	}
 	cmdutil.AddPodRunningTimeoutFlag(cmd, defaultPodPortForwardWaitTimeout)
+	cmdutil.AddAddressFlag(cmd)
```
prefer binding to the Address field in the struct, and defaulting to "127.0.0.1", "::1" above when constructing `opts` above

```
cmd.Flags().StringSliceVar(&opts.Addresses, "address", opts.Addresses, "Addresses to listen on (comma separated)")
```

liggitt commented on staging/src/k8s.io/client-go/tools/portforward/portforward.go:
```diff
@@ -184,10 +246,18 @@ func (pf *PortForwarder) forward() error {
-		return fmt.Errorf("All listeners failed to create with the following errors: %s, %s", errTcp4, errTcp6)
+	var listenSuccessfull bool
+	var errors []error
+	for _, addr := range pf.addresses {
```
silently failing on some of the inputs is not what I would expect. I'm unsure whether we want to support the user specifying multiple addresses to bind to... would we expect the command to exit if any were unsuccessful, or if all were unsuccessful? I could see both ways, which makes me think it might be better to limit a single invocation to a single address.

goblain replied:
at it's current implementation, it binds to two addresses (127.0.0.1, ::1) by default so that behavior should be kept, it is in my opinion perfectly logical to want to listen on particular set of IP addresses and not the others (ie. IP from one external interface and the loopback ones). The command in kubectl explicitly lists ports it successfully bound to

liggitt replied:
> at it's current implementation, it binds to two addresses (127.0.0.1, ::1) by default so that behavior should be kept

sure, if the user doesn't have an address they want to forward on, it should continue binding to "local" addresses.

> it is in my opinion perfectly logical to want to listen on particular set of IP addresses and not the others (ie. IP from one external interface and the loopback ones)

I agree, but if I explicitly want to listen on a set of IP addresses, I would not expect the command to report success if it couldn't accomplish that. Limiting the user-specified bind address to a single address makes it super clear when their intent was not satisfied, and plays well with something using this to script port-forward setups (e.g. loops that restart port-forward on a particular address when it times out or dies)

liggitt replied:
> The argument for scripting "compatibility" makes a lot of sense. I would lean to the idea that it could be implemented with failing if any bind failed

There's the initial bind failure to consider, but also what happens when one of the connections closes, right? I'd expect the following:
* On start, all explicitly requested bind addresses must succeed
* On close of any explicitly requested bind addresses, the entire command should exit

I would stick with the user specifying addresses explicitly (not doing localhost->[127.0.0.1, ::1] expansion for explicit addresses). You can move the defaulting into the command if it makes it easier to distinguish between unspecified addresses (which should preserve best-effort binding to 127.0.0.1 and ::1) and explicitly specified addresses of `127.0.0.1,::1` (which should both succeed or it would be considered failure)

goblain replied:
I was looking into it a bit more, and it seems in fact that port-forward is not really built to behave like this at the moment. Look at following vanilla kubectl port-forward : 

    $ kubectl port-forward mysql-server-0 3306:3306 80:3306  
    Forwarding from 127.0.0.1:3306 -> 3306
    Unable to listen on port 80: All listeners failed to create with the following errors: Unable to create listener: Error listen tcp4 127.0.0.1:80: bind: permission denied, Unable to create listener: Error listen tcp6: address [[::1]]:80: missing port in address

So... as long as one of the ports binds correctly, the port-forward stays up and running. 

I'm not saying that is the ultimate argument for not failing if only some of the addresses fail to bind, but to stay consistent, it would stand to reason that we change to fail if any port failed as well, which would be a change from current behavior.

lavalamp replied:
What do you think about changing it to just offer a single address? Users can always run multiple instances of kubectl, and that way they if one fails the user can restart it.

lavalamp replied:
I see there's a good case for running on all loopback addresses, at least.

In that case, I think it's super important that if it can't bind every address requested, it fail-- if something malicious is occupying one of the ports, it'd be a security vuln.

goblain replied:
as mentioned before, one of the cases where we will fail on one, but not all, is when we make a `localhost` bind on a system that does not support IPv6

lavalamp replied:
Since this PR isn't making things worse, maybe we should merge this and file an issue for followup.

liggitt commented:
> @smarterclayton:
> Yes, we're going to have to introduce a new kubelet to master to client
protocol for web sockets that can handle stream framing in order to handle
multiple connections being forwarded. I'd like to avoid anything impactful
to port forward until we make that change.

Clayton, what were the current limitations around the port-forward pieces? was it an inability to do multiple concurrent requests? exposing this on non-local addresses seems likely to increase usage impacted by that limitation. Would it make sense to come up with a clear description of the limitation and likely future impact of changes, and mark this flag as experimental so that people who want to use it now are aware of the status and future compatibility guarantees?

liggitt commented on pkg/kubectl/cmd/util/helpers.go:
```diff
@@ -394,6 +394,10 @@ func GetPodRunningTimeoutFlag(cmd *cobra.Command) (time.Duration, error) {
 	return timeout, nil
 }
 
+func GetAddressFlag(cmd *cobra.Command) []string {
```
these aren't general purpose... remove in favor of direct binding to the opts struct field

goblain replied:
sure, in fact it originally was direct, but the changes seemed like now this would be the preferred way. Anyway, will revert to previous way.

goblain commented:
@liggitt how about now ?

lavalamp commented on pkg/kubectl/cmd/portforward.go:
```diff
@@ -49,6 +49,7 @@ type PortForwardOptions struct {
 	RESTClient    *restclient.RESTClient
 	Config        *restclient.Config
 	PodClient     coreclient.PodsGetter
+	Address       []string
```
I don't know if this is still active or not, but it showed up in my email and since I glanced at it, surely this should be called "Interfaces", no?

goblain replied:
/test pull-kubernetes-integration

goblain replied:
As I understand, interfaces would be like `eth0` `wlp3s0` and similar, this is list of IP addresses to listen on

inc0 replied:
I agree with @goblain, address != interface and we're dealing with addresses here. Interface can have multiple addresses, we still need to specify which exactly we want to listen on.

goblain commented:
/test pull-kubernetes-integration

davidovich commented:
We are also very interested in this feature. What we need to do at the moment is a bit convoluted:

We have a development use-case where we want to share access to our minikube nginx-ingress-controller. Because external connections are not accepted, we are about to script this workaroud:

Start a kubectl proxy on the host running minikube (call it machine A):
```
$ kubectl config set-context minikube
$ kubectl proxy --accept-hosts='.*' --address='0.0.0.0'
```

On the remote machine (machine B), set a kubectl config context to allow access to machine A:

```
$ kubectl config set-cluster machineA --server=http://machineA.corporate.domain:8001 --insecure-skip-tls-verify=true
$ kubectl config set-context machineA-context --cluster=machineA
$ kubectl config use-context machineA-context
$ kubectl --namespace=kube-system port-forward nginx-ingress-controller-2wxt6 8080
```

Then enjoy access to machineA through machineB http://localhost:8080

If `kubectl port-forward` was allowed bind to external traffic, this would be reduced to a simple `kubectl port-forward --address='0.0.0.0'` call. This would also be inline with kubectl proxy.

goblain commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

goblain commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

goblain commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

rcoup commented:
> What's stopping you currently from invoking kubectl from the host you want the port be available?

For me: credentials and cluster context configuration being copied/passed around unnecessarily. Use case is connecting development VMs, dev runtimes, and dev tools to local (Docker/minikube) + remote k8s clusters

> If you need it, it looks like the built-in unix tools should do that for you, don't they? 

I'm sure socat _can_ do it if you cast the right spells and make the correct sacrifice, but it's not exactly intuitive. Especially to make it listen on everything-other-than-localhost and connect _to_ localhost.

> if this should be implemented I'd imagine one address being specified similarly to how you do that with tools known to me (including docker run), and that would be part of the `port:port`, as in `address:local_port:remote_port`, where address would be optional and defaulted to localhost.

👍

dims commented:
cc @thockin 

Tim, thoughts please!

alluneediswill commented:
I would very much look forward to this feature as well. But for anyone who also lands on this page and in need of a solution, here is my workaround with the aforementioned socat
```
kubectl port-forward svc/${SERVICE} 8081:8080 > kubectl_forward.log 2>&1 &
socat TCP4-LISTEN:8080,fork TCP4:localhost:8081
```
It will forward all the traffic to port 8080 from this host to a service in a remote cluster at port 8080.

vladimirvivien commented:
This is great way to quickly test/validate apps running in pods.  Is there anything that can be done to see this through rather than let it rot?  

Maybe instead of introducing the ability to bind to multiple addresses, maybe keep it simple and just use one address for each invocation of port-forward:

```shell
kubectl port-forward --address 10.19.21.23 pod/mypod 8888:5000
```
Or as suggested above:

```shell
kubectl port-forward  pod/mypod 10.19.21.23:8888:5000
```
Would this (specify 1 address) get it this merged?  

Judging by the number of issues opened, this is something that is useful.  If there is a technical case against doing this at all, it should be stated.

raravena80 commented:
+1

Jeeppler commented:
+1

svenkost commented:
+1 I really need this feature.

jason008bourne commented:
+1 I really need this feature.

goblain commented:
@mengqiy @liggitt @smarterclayton : what's the status for this? should I rebase in hopes it can be merged anytime soon, or is it still in cold storage for undefined time?

smarterclayton commented:
We can remove the hold.  The other stuff got delayed, and this won't make it worse.

goblain committed:
port-forward listen on address

adds an address flag to kubectl port-forward to allow to bind to a different ip then localhost

goblain commented:
rebased, chances to get lgtm here ?

dims commented:
/lgtm

/assign @liggitt

goblain commented:
/test pull-kubernetes-integration

goblain commented:
/test pull-kubernetes-e2e-kops-aws

goblain commented:
rebased yet again

@dims @mengqiy @liggitt @soltysh chances to get final review and approval / merge before another rebase is required ? :)

soltysh reviewed (approved):
/lgtm
/approve

dixudx reviewed (commented):
Actually I was wondering whether we could use `10.19.21.23:8888:5000` to indicate port 8888 open on 10.19.21.23 and forward to 5000 in the pod.

lavalamp commented on pkg/kubectl/cmd/portforward/portforward.go:
```diff
@@ -113,6 +120,7 @@ func NewCmdPortForward(f cmdutil.Factory, streams genericclioptions.IOStreams) *
 		},
 	}
 	cmdutil.AddPodRunningTimeoutFlag(cmd, defaultPodPortForwardWaitTimeout)
+	cmd.Flags().StringSliceVar(&opts.Address, "address", []string{"localhost"}, "Addresses to listen on (comma separated)")
```
I suggest "local-addresses"?

goblain replied:
`address` was selected as it's the same as for ie. `kubectl proxy`, so makes sense to stay coherent

lavalamp commented on pkg/kubectl/cmd/portforward/portforward.go:
```diff
@@ -50,6 +50,7 @@ type PortForwardOptions struct {
 	RESTClient    *restclient.RESTClient
 	Config        *restclient.Config
 	PodClient     corev1client.PodsGetter
+	Address       []string
```
LocalAddresses?

lavalamp replied:
(but see my suggestion to only serve one local address)

lavalamp commented on staging/src/k8s.io/client-go/tools/portforward/portforward.go:
```diff
@@ -110,8 +111,52 @@ func parsePorts(ports []string) ([]ForwardedPort, error) {
+	var addresses []listenAddress
+	parsed := make(map[string]listenAddress)
+	for _, address := range addressesToParse {
+		if address == "localhost" {
```
This looks wrong. I think you need to resolve the name rather than hard coding this stuff.

lavalamp replied:
Likely you should attempt resolving any name that isn't an IP address.

lavalamp commented on staging/src/k8s.io/client-go/tools/portforward/portforward.go:
```diff
@@ -110,8 +111,52 @@ func parsePorts(ports []string) ([]ForwardedPort, error) {
+
+// New creates a new PortForwarder with localhost listen addresses.
 func New(dialer httpstream.Dialer, ports []string, stopChan <-chan struct{}, readyChan chan struct{}, out, errOut io.Writer) (*PortForwarder, error) {
+	return NewOnAddresses(dialer, []string{"localhost"}, ports, stopChan, readyChan, out, errOut)
```
Perhaps explicitly say 127.0.0.1 & ::1 here, then you can remove the special case in the parse function?

lavalamp replied:
In a followup, I think we should do this after confirm that the address exists, if possible.

lavalamp commented on staging/src/k8s.io/client-go/tools/portforward/portforward.go:
```diff
@@ -110,8 +111,52 @@ func parsePorts(ports []string) ([]ForwardedPort, error) {
+	return NewOnAddresses(dialer, []string{"localhost"}, ports, stopChan, readyChan, out, errOut)
+}
+
+// NewOnAddresses creates a new PortForwarder with custom listen addresses.
```
Document that addresses must all be valid IPv4 or IPv6 addresses, no hostnames?

lavalamp commented:
/approve

I think other changes can be made in a followup.

vladimirvivien commented:
Wow.  This is awesome.

Dimagog commented:
This is great!
**Where can I find Windows binary with this change?**
I need it badly and all my attempts to build kubectl myself fail with:
```
E:\Go\home\src\k8s.io\kubernetes\pkg\features\kube_features.go:458:23: cannot use "k8s.io/apiextensions-apiserver/pkg/features".CustomResourceValidation (type "k8s.io/apiextensions-apiserver/vendor/k8s.
io/apiserver/pkg/util/feature".Feature) as type "k8s.io/apiserver/pkg/util/feature".Feature in map key
E:\Go\home\src\k8s.io\kubernetes\pkg\features\kube_features.go:459:23: cannot use "k8s.io/apiextensions-apiserver/pkg/features".CustomResourceSubresources (type "k8s.io/apiextensions-apiserver/vendor/k8
s.io/apiserver/pkg/util/feature".Feature) as type "k8s.io/apiserver/pkg/util/feature".Feature in map key
```

mengqiy commented:
@Dimagog it should be in this prerelease: https://github.com/kubernetes/kubernetes/releases/tag/v1.13.0-alpha.1

Dimagog commented:
@mengqiy Yes, latest binary from below has the `--address` switch, thanks!
https://storage.googleapis.com/kubernetes-release/release/v1.13.0-alpha.1/bin/windows/amd64/kubectl.exe
